### DISCUSSION TITLE
AWS Systems Manager Parameter store support for Airflow AWS Provider package

### DISCUSSION THREAD

**Discussion by hrishik** [apache/airflow#21012] [General] [answered]
### Description

Airflow DAGs when used in AWS context, the parameter sharing across different subsystems becomes tedious job. There is a need to create a central parameter handling repository that can be accessed by any service provider (Such as AWS Service) outside of Airflow DAG. 

### Use case/motivation

One important part of DAG based orchestration is sharing of metadata and configuration across different service providers.
Existing alternative of using XCom feature of Airflow has its own limitations in terms of security, encryption, extensibility etc., 

AWS Systems Manager parameter store provides the parameter sharing mechanism, and it is directly supported by all AWS services including lambda, glue, cloud-formation etc. It is an important addition to Airflow to develop the operator for AWS Systems Manager parameter store to standardize information sharing across different services and Airflow DAG seamlessly. 

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
The SystemsManagerParameterStoreBackend is already available in Airflow as custom Secrets manager: 

https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_api/airflow/providers/amazon/aws/secrets/systems_manager/index.html#airflow.providers.amazon.aws.secrets.systems_manager.SystemsManagerParameterStoreBackend

**Comment by hrishik** [comment]
> The SystemsManagerParameterStoreBackend is already available in Airflow as custom Secrets manager:
> 
> https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_api/airflow/providers/amazon/aws/secrets/systems_manager/index.html#airflow.providers.amazon.aws.secrets.systems_manager.SystemsManagerParameterStoreBackend

The existing operator only provides a read for specific actions i.e. connection URI, Config, and custom variable pertaining to airflow information. I am talking about using SSM parameter store as metadata storage to share configuration/metadata across different subsystems, as well as airflow. This means all actions including setting and getting parameters.

**Comment by potiuk** [comment]
I don't thik Airlflow will never have "AWS" or "Google" or any other "specific" configuration built in in other way that current "customization options".  Airflow is "Cloud Agnostic". 

It provides features to "talk" to external systems (via Providers: Hoops, Operators, Sensors) and airflow configuration capabilities can be extended precisely via Custom Secrets Backend. 

If you also want to use SystemsManagerParameter for sharing Inter-tasks metadata - feel free to add and contribute custom XCom Backend using  https://airflow.apache.org/docs/apache-airflow/stable/concepts/xcoms.html#custom-backends in the Amazon Provider. But that does not require a separate issue or agreement, just contribute it to the AWS provider.

**Comment by potiuk** [comment]
BTW. I converted it to discussion. If you feel the answer you got is not enough and you want to have something else - feel free to elaborate here. If you think Custom Secret Backend, Custom XCom is not enough - please describe what you mean by 

> I am talking about using SSM parameter store as metadata storage to share configuration/metadata across different subsystems, as well as airflow. This means all actions including setting and getting parameters

Ideally showing examples of how you think the configuration and DAGs should be affected by the proposal and why exactly current customization options are not good enough (ideally a specific example where you show where the current customization options will break).

**Reply by potiuk** [reply]
BTW. Nice guide about writing a custom XCom backend is here: https://www.astronomer.io/guides/custom-xcom-backends please take a closer look and read it in detail and provide the examples of how you would like to see differet integration with "Systems Manager Parameter Store". I hope I will be able to convince you that what you want to do is already possible today.

**Reply by schirag1993** [reply]
Thank you for opening this discussion @potiuk. I've responded below with more details of a use case I have for this feature.

**Comment by schirag1993** [comment, marked as answer]
Hello @potiuk,

Customers often use AWS Systems Manager Parameter Store to retain parameters that are not necessarily required only for Airflow DAG configurations.

We have use cases where we create a baseline configuration for VMs (such as required packages, OS type, vCPU and memory configurations) in an Airflow DAG and then store this baseline VM configuration information in AWS parameter store to be retrieved by our centrally managed deployment system (AWS CloudFormation based).

This allows VM config info to be available to users/entities that do not have access to our Airflow UI.

While we do not directly use AWS Systems Manager Parameter Store to "configure" our DAG runs, it is being used to communicate/share DAG outputs to other entities within our larger system. I believe this does not break the "Cloud Agnostic" tenet. Rather, it provides another way for Airflow to interact/talk to Cloud providers (in this case AWS) via operators, hooks, sensors, etc.

Would this serve as a viable use case to adding this feature into the AWS Provider package?

**Reply by potiuk** [reply]
I still do not understand what you actually propose. 

Can you provide example on how such interaction would look like from Airflow/DAG perspective if implemented?

**Reply by potiuk** [reply]
For me it looks like you want to update some of the operators from AWS provider to use System Manager Parameter as configuration, which  is a valid use case and can be implemented operator-by-operator, but it does not require any changes to Airflow core, discussion nor issue - just PR with the changes will be enough. But maye I do not understand what you propose, so please provide an example.

**Reply by schirag1993** [reply]
Yes, please allow me to elaborate further:

Problem statement: Maintain and manage a configuration for our VMs that adhere to all of our organization's security best practices and developer requirements (vCPU count, memory size, OS distro, required (and approved) packages, etc.) that can be used to deploy new VMs (EC2 instances) via CloudFormation. 

Implementation: 
An Airflow DAG (runs weekly) that:
1. starts an EC2 instance aka VM (via [EC2StartInstanceOperator](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_api/airflow/providers/amazon/aws/operators/ec2/index.html#module-airflow.providers.amazon.aws.operators.ec2))
2. Installs relevant packages (via [ssh_operator](https://airflow.apache.org/docs/apache-airflow/1.10.12/_api/airflow/contrib/operators/ssh_operator/index.html)) 
3. Runs test scripts to verify proper functioning of VM and if successful
4. Stores updated configuration details (such as package versions, OS distro version aka VM image identifier, AD join details, etc.) into a parameter in parameter store as output of DAG. 

This verified configuration acts as the baseline for any VM that is launched in our account through our deployment system (which is CloudFormation + a few other Cloud services). Since our deployment system is not aware of Airflow, there is no way of storing this configuration for other AWS entities to use (in this case CloudFormation). 

References for more details around AWS Systems Manager parameter store:
[Documentation](https://docs.aws.amazon.com/systems-manager/latest/userguide/systems-manager-parameter-store.html)
[High level use cases](https://docs.aws.amazon.com/systems-manager/latest/userguide/systems-manager-best-practices.html)

**Reply by schirag1993** [reply]
And yes, you are right: We do want to implement operator-by-operator. We do not want to make a change to Airflow core. We simply want to submit a PR to create a new pair of operators.

**Reply by potiuk** [reply]
Feel free to submit them then ! I thought you wanted to make some changes to "Airlfow Core" with those. 

FYI. We are not fans of beurocraucy. Our changelogs and releases are based on PRs not on issues. Most of changes like that - especially to providers, when you add new operators or want to augment existing ones  - do not require opening issues or heavy discussion. Just open  PR with proposed change and the discussion can carry on there.

**Reply by schirag1993** [reply]
Awesome! Thank you for the healthy discussion @potiuk. We'll submit our contribution based on this guidance moving forward.
